include/aws/crt/endpoints/RuleEngine.h

```diff
@@ -0,0 +1,157 @@
+             * previous values.
+             */
+            class AWS_CRT_CPP_API RequestContext final
+            {
+              public:
+                RequestContext(Allocator *allocator = ApiAllocator()) noexcept;
+                ~RequestContext();
+
+                /* TODO: move/copy semantics */
+                RequestContext(const RequestContext &) = delete;
+                RequestContext &operator=(const RequestContext &) = delete;
+                RequestContext(RequestContext &&) = delete;
+                RequestContext &operator=(RequestContext &&) = delete;
+
+                /**
+                 * @return true if the instance is in a valid state, false otherwise.
+                 */
+                operator bool() const noexcept { return m_requestContext != nullptr; }
+
+                /*
+                 * Add string parameter.
+                 * True if added successfully and false if failed.
+                 * Aws::Crt::LastError() can be used to retrieve failure error code.
+                 */
+                bool AddString(const ByteCursor &name, const ByteCursor &value);
+
+                /*
+                 * Add boolean parameter.
+                 * True if added successfully and false if failed.
+                 * Aws::Crt::LastError() can be used to retrieve failure error code.
+                 */
+                bool AddBoolean(const ByteCursor &name, bool value);
+
+                /// @private
+                aws_endpoints_request_context *GetNativeHandle() const noexcept { return m_requestContext; }
+
+              private:
+                Allocator *m_allocator;
+                aws_endpoints_request_context *m_requestContext;
+            };
+
+            /*
+             * Outcome of Endpoint Resolution.
+             * Outcome can be either endpoint (IsEndpoint) or error (IsError).
+             * Endpoint outcome means that engine was able to resolve context to
+             * an endpoint and outcome can have the following fields defined:
+             * - Url (required) - resolved url
+             * - Headers (optional) - additional headers to be included with request
+             * - Properties (optional) - custom list of properties associated
+             *   with request (json blob to be interpreted by the caller.)
+             *
+             * Error outcome means that context could not be resolved to an endpoint.
+             * Outcome will have following fields:
+             * - Error (required) - error message providing more info on why
+             *   endpoint could not be resolved.
+             */
+            class AWS_CRT_CPP_API ResolutionOutcome final
+            {
+              public:
+                ~ResolutionOutcome();
+
+                /* TODO: move/copy semantics */
+                ResolutionOutcome(const ResolutionOutcome &) = delete;
+                ResolutionOutcome &operator=(const ResolutionOutcome &) = delete;
```

Comment:
very minor nit
why & next to operator not the type?

Reply:
dont know the history of this, but this is the format used across the package, so just staying consistent